Practical productization enhancements

cbruiz opened the pull request:
Few changes proposed to facilitate integration in embedded projects.
Most important could be the ability to process gcode words with rust match templates almost easily.

* Use lightweight fixed decimal representation to facilitate matching and avoid precision issues.
* Internal (num, ord) decimal logic reinterpretation.
* Make f64 optional so it could be non fully/completely supported in baremetal.

cbruiz committed:
Practical productization enhancements
* Use lightweight fixed decimal representation to faciliate matching and avoid precision issues.
* Internal (num, ord) decimal logic reinterpretation.
* Make f64 optional so it could be non fully/completely supported in baremetal.

ithinuel commented on src/types.rs:
```diff
@@ -52,19 +122,26 @@ impl Literal {
-        Self::RealNumber(from)
+        Self::RealNumber(DecimalRepr::from_f64(from))
     }
 }
```
Those conversions are lossy, so they should be `TryFrom` instead.

ithinuel commented on src/parser/values.rs:
```diff
@@ -91,21 +85,21 @@ where
-    //);
+    #[cfg(feature = "defmt")]
+    info!("parse_real_literal_alt: int: {}, dec: {}", int, dec);
 
```
This whole bit could be even removed TBH. It's a left over of debugging sessions :D

ithinuel commented on src/parser/values.rs:
```diff
@@ -12,35 +12,31 @@ use futures::{Stream, StreamExt};
         let b = match input.next().await? {
             Ok(b) => b,
-            Err(e) => return Some(Err(e)),
+            Err(e) => {
+                return Some(Err(e))},
```
Is this a result of rustfmt ?

cbruiz replied:
Not sure what you mean. It's a result of one of the parser. I splitted in lines, but there is no change there

ithinuel commented on .gitignore:
```diff
@@ -5,3 +5,4 @@ kcov*
 tags
 **/*.rustfmt
 Cargo.lock
+.idea
```
I'd rather keep IDE specific things outside of the repository's config (eg I use vim rather than jetbrains or vscode).
You can set a system wide gitignore, see: https://blog.sebastian-daschner.com/entries/global-gitignore

ithinuel reviewed (changes requested):
Thank you for the contribution.

Removing the requirement for `f64` would be welcome indeed!
However, I am not sure that what is in this PR really makes a positive impact on the generated binary.

I had initially implemented it using binary scaling fixed point arithmetic.
It is a shame I got rid of the branches with the initial implementation 🤦 

At the time, I ran tiny benchmark to compare the generated output binary.
Using the ATSAM4E (Cortex-M4 with single precision FPU), and keeping a position accuracy to 10^-5 mm which is the accuracy used by PrusaSlicer for the extruder axis. That is 17bits for the fractional part, 14bits for the integer part (on a signed integer) to fit a 32bits variable.
This limited the length to about 16meters. Although it may look fine for X,Y and Z, the slicer may emit an E position that's beyond that limit.

Implementing math for these (eg using https://crates.io/crates/fixed with its `FixedI32<U17>`) was increasing the binary size by about as much as what the double-precision float (which is also software).

So I figured that f64 was providing better precision (53bits mantissa ~ 15.95 decimal digits) at the expanse of some ram to store 64bits rather than 32.

The representation proposed in this PR effectively also takes 64bits (after accounting for padding & alignment) but does not implement the common math operations.

So here are a few questions that I'd like answered before proceeding:
- Do you have figure to sustaining that this is effectively lighter weight (after considering math/conversion etc)?
- This DecimalRepr seems to me like a manual version of floating point (32bit signed mantissa + 8bit unsigned? exponent). What is the benefit over a f64 and/or, say, a `FixedI64<U17>` fixed point representation?

cbruiz committed:
Merge review

cbruiz commented:
> So here are a few questions that I'd like answered before proceeding:
> 
> * Do you have figure to sustaining that this is effectively lighter weight (after considering math/conversion etc)?

My need was related to the comparison of gcodes in a higher level in order to facilitate matching by constant avoiding precision deviations. 
> * This DecimalRepr seems to me like a manual version of floating point (32bit signed mantissa + 8bit unsigned? exponent). What is the benefit over a f64 and/or, say, a `FixedI64<U17>` fixed point representation?

Yes. It's manual. I didn't wanted to put a constraint by selecting one of the many crates providing fixed point or just use f64 by activating the feature. I think is more practical. The manual representation implemented is close to rust_decimal.

rust_decimal has HUGE precision and memory cost (more than I need), but also provides several math operations I needed... at least to evaluate. Nevertheless, as said.. maybe is good to let the consumer decide.

cbruiz committed:
Control excess of precision while parsing to avoid overflows

cbruiz committed:
- Experiment: Add an extra simple interface leveraging async-traits
- Add support to collect and reestablish line counting

cbruiz committed:
Allow negative line numbers for some gcode producers requesting for instance: "N-1 M110 N-1" (Printrun)
Fix wrong signature from previous commit

cbruiz committed:
Experimental recovery check

cbruiz committed:
Experimental support for string values without word of quotes to support (relaxing standard), for instance M118 Hello World

cbruiz committed:
Fix feature issue when needed string-value without parse-checksum after the "string values without word of quotes to support (relaxing standard)" implementation applied
Fix pending issue with missing recovery_check method after such capability was implemented
